cmd/antrea-agent/options.go

```diff
@@ -82,6 +89,34 @@ type Options struct {
 	nplEndPort             int
 	dnsServerOverride      string
 	nodeType               config.NodeType
+
+	ovsDatapathType         ovsconfig.OVSDatapathType
+	tunnelType              ovsconfig.TunnelType
+	trafficEncapMode        config.TrafficEncapModeType
+	trafficEncryptionMode   config.TrafficEncryptionModeType
+	ipsecAuthenticationMode config.IPsecAuthenticationMode
+	transportInterfaceCIDRs []*net.IPNet
+
+	enableAntreaProxy bool
+	nodePortAddresses []*net.IPNet
+
+	enableAntreaPolicy    bool
+	enableL7NetworkPolicy bool
+
+	enableEgress bool
+	exceptCIDRs  []*net.IPNet
+
+	enableMulticast bool
+
+	enableTrafficControl    bool
+	enableServiceExternalIP bool
+	enableAntreaIPAM        bool
+	enableSecondaryNetwork  bool
+
+	serviceCIDR   *net.IPNet
+	serviceCIDRv6 *net.IPNet
+
+	policyBypassRules []*types.PolicyBypassRule
```

Comment:
I do not see all config parameters (e.g. TransportInterface, NodePortLocal) and all featuregates (e.g. Multicluster) in `Options`. What is the criterial to decide what parameters should be added to `Options`?

Reply:
As the comment of the first such field explains, "The following fields contains the configuration values that have been validated and converted to Go types from the configuration file."

Actually I created a separate object to contain all validated and converted configuration values, and only injected the validated config object to `run` function only, like the following:
```
type Options struct {
	// configFile is the path of configuration file.
	configFile string
	// config is the configuration object parsed from configuration file.
	config *agentconfig.AgentConfig
	// CompletedConfig is the validated configuration converted from the configuration object.
	*CompletedConfig
}

// CompletedConfig contains the configuration options that have been validated and converted to Go types from the
// configuration file. It's the struct consumed by the run function of antrea-agent.
// For any user provided configuration value, it should be validated, optionally converted to Go types, and stored in
// this struct.
type CompletedConfig struct {
	NodeType config.NodeType

	// CNI related configurations.
	CNISocket                string
	HostGateway              string
	DefaultMTU               int
	HostProcPathPrefix       string
	DisableTXChecksumOffload bool
	TransportInterface       string
	TransportInterfaceCIDRs  []*net.IPNet

	// Client related configurations.
	ClientConnection       componentbaseconfig.ClientConnectionConfiguration
	AntreaClientConnection componentbaseconfig.ClientConnectionConfiguration
	KubeAPIServerOverride  string
        ...
	EnableAntreaIPAM        bool
	EnableBridgingMode      bool
	EnableSecondaryNetwork  bool
	EnableServiceExternalIP bool
	EnableTrafficControl    bool
        ...
}

if err := opts.complete(args); err != nil {
	klog.Fatalf("Failed to complete: %v", err)
}
if err := opts.validate(args); err != nil {
	klog.Fatalf("Failed to validate: %v", err)
}
if err := run(opts.CompletedConfig); err != nil {
	klog.Fatalf("Error running agent: %v", err)
}
```

The considerations are:
1. All user inputs should be validated to avoid illegal input and securty problem.
2. The program (`run` func) should consume validated configuration only (the `CompletedConfig`) and shouldn't even need the access of the config parsed from user input (the `agentconfig.Config`), which could force a validation when new parameter is added, as there is no direct access to user input.
3. Avoid unexpected runs of unapplicable modules, every optional module will be controlled by a `enableX` flag in`CompletedConfig` and defaults to false and will only be set to true if its validation pass. If a scenario (ExternalNode NodeType) doesn't even call the validation function, the module will ensure not run.

But I found the size of the patch is getting larger so reverted some of them. Do you think the above approach makes more sense. If you also agree, I could add that change back for all these consideration.

Reply:
Still trying to understand. So, you mean only if we convert e.g. from a string type raw value to a Go type, we then add field?

I do not understand the "validated" point. All Options.config fields are validated, otherwise Options.validate() will fail?

For `enableX`, should we add one for every featuregate? Or how we decide what should be an `enableX` field?

Reply:
> Still trying to understand. So, you mean only if we convert e.g. from a string type raw value to a Go type, we then add field?

In current PR, yes. But I propose to have a separate struct for configurations, including the ones converted from user input and the ones deduced from user input and platform (like EnableEgress). You can see it as an internal config converted from user config.

> 
> I do not understand the "validated" point. All Options.config fields are validated, otherwise Options.validate() will fail?

Currently `run` have direct access to user provided configuration object, so actually many inputs that should have been validated are not validated and are consumed by `run` directly, e.g. various CIDR inputs, string inputs.

By having an internal config and only pass it to `run`, it makes it mandatory that all user provided must be converted to fields in internal config, and normally the process of conversion contains invalidation, like from cidr string to `net.IPNet`, and even for a simple string representing some resource's name, like ExternalNodeNamespace, we should validate it matches K8s resource's format, and copy it to internal config only if it passes.

> 
> For `enableX`, should we add one for every featuregate? Or how we decide what should be an `enableX` field?

It depends on whether X is optional. If it's optional in any scenario, there should be a `enableX` flag, and we calculate it based on user input and platform in validation phase.

Reply:
Thanks for the explanation. Clearer to me now. I feel an internal struct, uniformly for all config parameters is easier to understand. Another thing to think about is the relation with other config struct, like NetworkConfig, NodeConfig (which include values from both config parameters and discovered state like Node IPs).